File: 6th_Semester/NET_Centric_Computing/Class codes/Basics/Exception_Handling.cs
/*
 * Exception is any kind of error that disturbs the normal flow of the program.
 * if exception occurs, then whole program is stopped.
 * to deal with such error exception handling mechanism is required.
 * it contains 4 keywords
 *
 * 1. try: use to surround a code that can cause error
 * 2. catch: use to handle the error
 * 3. throw: use to throw an exception
 * 4. finally: this block will always execute whether exception occurs or not
*/
using System;

namespace Basics
{
    class ExceptionHandling
    {
        private int x;
        private int y;

        public void SetData()
        {
            Console.WriteLine("Enter first number: ");
            x = Convert.ToInt32(Console.ReadLine());

            Console.WriteLine("Enter second number: ");
            y = Convert.ToInt32(Console.ReadLine());
        }

        public void Divide()
        {
            try
            {
                int div = this.x / this.y;
                Console.WriteLine("Division: " + div);
            }
            catch (DivideByZeroException d)
            {
                Console.WriteLine($"Cannot divide {this.x} by {this.y}.\n {d}");
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }
            finally
            {
                Console.WriteLine("This block will always execute whether exception occurs or not.");
            }
            Console.WriteLine("Out of try/catch block.");
        }
    }
}

File: 7th_Semester/Advanced Java Programming/Class Codes/src/main/java/Unit1/Exception/ExceptionHandling.java
package Unit1.Exception;

import java.util.Scanner;

/**
 * Exception handling: It is a mechanism to handle runtime errors so that normal flow of the application can be maintained
 * <p>
 * try: any code section that can produce exception is placed on try block
 * catch: if any error occur in try block then such exception should be caught in catch block
 * throw: to explicitly throw an exception
 * throws: if exception can occur from method then it should be declared in method signature
 * finally: block that will always execute whether exception occurs or not
 * <p>
 * difference between throw and throws:
 * 1) throw: used to throw an exception explicitly
 * 2) throws: used to declare an exception
 */
public class ExceptionHandling {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int a = sc.nextInt();
        int b = sc.nextInt();

        // if value of b is 0, then exception will occur, which is handled by catch block
        try {
            int c = a / b;
            System.out.println("Division is: " + c);

            // ArrayIndexOutOfBoundsException occurs when array index is out of bound
            int arr[] = new int[2];
            arr[0] = 1;
            arr[1] = 2;
            arr[2] = 3;
        } catch (ArithmeticException | ArrayIndexOutOfBoundsException e) {
            System.out.println("Exception occurred: " + e);
        } catch (Exception e) {
            System.out.println("Exception occurred: " + e);
        } finally {
            System.out.println("Finally block executed");
        }

        System.out.println("Out of try-catch block");
    }
}
